devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers of data call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers of data call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## `waiter` state
+
+The state of the `waiter` module consists of three components:
+- `EPOCH`: This is a monotonically increasing integer that is incremented every time an operation is completed or cancelled.
+- `NOTIFIED`: This is a bit signifying that `waiter_notify` has been called previously and a `waiter_register_notification` call is expected to complete the operation.
+- `REGISTERED`: This is a bit signifying that `waiter_register_notification` has been called previously and a `waiter_notify` call is expected to complete the operation.
+
+All three components are stored in a single `volatile_atomic` variable that is manipulated using `interlocked` APIs.
+
+```mermaid
+stateDiagram-v2
+    B: EPOCH=x,N=0,R=0
+    N: EPOCH=x,N=1,R=0
+    R: EPOCH=x,N=0,R=1
+    E: EPOCH=x+1,N=0,R=0
+
+    B --> N:  waiter_notify
+    B --> R:  waiter_register_notification
+
+    N --> E:  waiter_register_notification
+    R --> E: waiter_notify
+
+    N --> E:  cancel waiter_notify
+    R --> E: cancel waiter_register_notification
+
+    R --> R: waiter_register_notification
+    N --> N: waiter_notify
+```
+
+## Reentrancy
+
+`waiter` is reentrant. This means that `waiter_register_notification` and `waiter_notify` can be called from `NOTIFICATION_CALLBACK` and `NOTIFY_COMPLETE_CALLBACK` functions.
+
+## Exposed API
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_SYNC, \
+    WAITER_RESULT_ASYNC, \
+    WAITER_RESULT_REFUSED, \
+    WAITER_RESULT_ERROR
+
+MU_DEFINE_ENUM(WAITER_RESULT, WAITER_RESULT_VALUES);
+
+#define WAITER_CALLBACK_RESULT_VALUES \
+    WAITER_CALLBACK_RESULT_OK, \
+    WAITER_CALLBACK_RESULT_CANCELLED, \
+    WAITER_CALLBACK_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_CALLBACK_RESULT, WAITER_CALLBACK_RESULT_VALUES);
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+typedef void(*NOTIFICATION_CALLBACK)(void* context, THANDLE(RC_PTR) data, WAITER_CALLBACK_RESULT result);
+typedef void(*NOTIFY_COMPLETE_CALLBACK)(void* context, WAITER_CALLBACK_RESULT result);
+typedef struct WAITER_TAG WAITER;
+
+THANDLE_TYPE_DECLARE(WAITER);
+
+    MOCKABLE_FUNCTION(, THANDLE(WAITER), waiter_create)
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_register_notification, THANDLE(WAITER), waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
```

Comment:
>out_op [](http://example.com/codeflow?start=176&length=6)

just so they don't get confused, consider maybe calling this "out_op_consumer" and the other one "out_op_producer"